### ISSUE TITLE
F# compilation very slow with FCS v41.0.1

### ISSUE THREAD

**Issue by alfonsogarciacaro** [fable-compiler/Fable#2605]
I didn't notice any difference when compiling the tests, but after switching the custom FCS build with the v41.0.1 Nuget package and testing it with a big project (1275 files and I believe it heavily uses SRTP) I realized the F# compilation is very slow, it takes almost 2 times as long as with current Fable version (3.4.10, note that Fable with FCS Nuget hasn't been yet released:

```
Fable 3.4.10

F# compilation finished in 65379ms
Fable compilation finished in 15852ms

main branch (actually queue-watch)

F# compilation finished in 102581ms
Fable compilation finished in 17020ms
```

The F# compilation time only measures the [`ParseAndCheckProject` call](https://github.com/fable-compiler/Fable/blob/f9cfffcedd67324cacdbc10a4fdd82bfb8f22855/src/Fable.Cli/Main.fs#L340-L343) so the performance decrease must come from FCS, unless I'm initializing the options or the checker in a  wrong way.

@ncave I guess it will be difficult to say with so little information, but do you have any idea where the problem may come from? Latest `main` is built with dotnet sdk 6 but targets net5.0 (I did try to target net6.0 but there was no difference). I remember you pruned off some features in `service_slim` but don't remember if this also impacted performance or only memory. Should I try to create a repro and report the issue in the dotnet/fsharp repo (problem is the project mentioned above is not public)?

**Comment by ncave** [comment]
@alfonsogarciacaro But you don't see any difference compiling Fable itself? Then it has to be some difference in FCS, perhaps can you try older versions of the FCS package to see if that's true.

**Comment by alfonsogarciacaro** [comment]
@ncave You were right, the performance decrease seems to have happened already in a previous release, specifically in v40.0.1-preview. I also tried building standalone but I keep getting [this error](https://github.com/fable-compiler/Fable/pull/2573#issuecomment-963784459) even with FCS v40.0, which maybe related to [the hack you had to make in your fork](https://github.com/ncave/fsharp/commit/0c690987a0cfaed1db6a2a3f9961c938e8c65263) for Witness issues. It's weird because I think at one point [it was working](https://github.com/fable-compiler/Fable/pull/2573#issuecomment-954629530) but maybe I made a mistake when testing.

At one point I also tried to create a custom build with latest dotnet/fsharp/main, but I'm getting errors about packages missing from custom Nuget feeds. I missed this brief moment when it was very easy to build FSC :)

We probably should defer the F# 6 upgrade until we can investigate this more, so going back to your fork for now :) 

I've reported this in dotnet/fsharp repo: https://github.com/dotnet/fsharp/issues/12380

**Comment by ncave** [comment]
@alfonsogarciacaro Can you use v40.0 for now?

**Comment by alfonsogarciacaro** [comment]
@ncave I was trying to reenable standalone again, but now I'm getting some errors related to how I'm [dealing with inline functions now](https://github.com/fable-compiler/Fable/commit/9045c154050d14460c735d2c9cd360e7d4f553cb#diff-1d58bcc2a844ad7919c9b2df626b85acec712971c11cc227e8bcdba717f3ee15R121) :) I'm trying to collect now the inline info in advance same as we do with entities. I should fix those anyways.

Interestingly, when using your fork and the "standard" FCS API I still get worse performance ([this commit](https://github.com/fable-compiler/Fable/commit/fe97ba1ddb03d53c3129c7ce9c9df61bf41903f2)):

```
Fable 3.4.10: F# compilation finished in 63421ms
Commit fe97ba1ddb03d53c3129c7ce9c9df61bf41903f2: F# compilation finished in 78840ms
```

So there's benefit in keeping the service_slim, but I need some fixes (like adding [SourceIdentifier here](https://github.com/ncave/fsharp/blob/d3fd0b7ebc9e3389ab1349bb3d4979e8da652697/src/fsharp/CheckExpressions.fs#L4418)) so now I've to make a fork of your fork :)

**Comment by ncave** [comment]
@alfonsogarciacaro Sounds like you're saying there may be still some benefits from having a facade on top of FCS. I'll try to find some time to look at updating the fork.

**Comment by alfonsogarciacaro** [comment]
@ncave Seems so. I had some success in [rebasing `service_slim`](https://github.com/alfonsogarciacaro/fsharp/tree/service_slim_2021-06-08) with dotnet/fsharp up to 2021-06-08 (which contains [this fix](https://github.com/dotnet/fsharp/commit/78826fdd4d5918eb4d42bf8f2799b7347d2aa60c)). I managed to build FCS by using [this command](https://github.com/alfonsogarciacaro/fsharp/blob/13c6c69746e143c6ba572f13b073fae1c7d4f5d4/fcs/build.sh#L5) and I also tried adding a [`lastFile` argument](https://github.com/alfonsogarciacaro/fsharp/blob/13c6c69746e143c6ba572f13b073fae1c7d4f5d4/src/fsharp/FSharp.Compiler.Service/service_slim.fs#L243-L253) to `ParseAndCheckProject`. I was hoping this would avoid the current worst case in watch compilation in Fable (touching a file up in the list triggers checking for all files below even if it's a leaf node) but I'm not getting the expected results (now touching the last file takes much longer than before). Still investigating :/

**Comment by ncave** [comment]
@alfonsogarciacaro I thought watch is now using `ParseAndCheckFileInProject`, no?

**Comment by dsyme** [comment]
If we can get any profiling information for comparison that would be great.

**Comment by alfonsogarciacaro** [comment]
@ncave Yes, but at the end `ParseAndCheckFileInProject` checks all files before the updated one, and because I know the bulk of changes in each watch iteration (the changed files plus the ones depending on them from Fable's [dependency tree](https://github.com/fable-compiler/Fable/blob/4ba8ba2dcf3b5b0634116c0840f48d2b9b72474a/src/Fable.Cli/Main.fs#L468-L479)), it's easier (and I assume it saves work) if I just tell FCS until what file it needs to recompile.

Thanks a lot for checking this @dsyme! I replied in the dotnet/fsharp issue :+1:

**Comment by alfonsogarciacaro** [comment]
Not entirely related, but some interesting data. With current `main` I get these timings:

* Compiling standalone (FCS + Fable itself)
    - F# compilation finished in 18105ms
    - Fable compilation finished in 22801ms

* Compiling big project (1275 files)
    - F# compilation finished in 64787ms
    - Fable compilation finished in 12285ms

So with standalone Fable compilation takes longer but with the big project is much faster (this is often the case). I wonder what's causing this difference, but I think this big project uses SRTP extensively (FCS resolves these for Fable now, so that would explain why it takes longer).

**Comment by ncave** [comment]
@alfonsogarciacaro Can you post the same numbers with Fable 3.4.10?

**Comment by alfonsogarciacaro** [comment]
@ncave Here you go:

* Compiling standalone (FCS + Fable itself) with Fable 3.4.10
    - F# compilation finished in 18844ms
    - Fable compilation finished in 25367ms

* Compiling big project (1275 files) with Fable 3.4.10
    - F# compilation finished in 61553ms
    - Fable compilation finished in 11920ms

They're very similar which makes sense because at the end `main` is using a version of FCS very close to 3.4.10. Also, as you can see [here](https://github.com/fable-compiler/Fable/issues/2605#issuecomment-966813988), FCS v40.0 from Nuget is still a bit slower that `service_slim` (I've updated the comment to include standalone timing).

**Comment by alfonsogarciacaro** [comment]
Some interesting things happening. I managed to [update ncave's `service_slim` to latest dotnet/fsharp](https://github.com/alfonsogarciacaro/fsharp/commits/service_slim_2021-11-12). See #2608 

I did the update in steps trying to find the point when performance dropped, but I couldn't find it. It's a bit slower with latest code which seems to correspond to F# 6 doing more thing but the difference is much smaller that when using FCS API from v41.0 from Nuget. Some timings:

```
2021-06-21 - Standalone
F# compilation finished in 16338ms
Fable compilation finished in 19487ms

2021-07-03 - Standalone
FCS: Parsing finished in 1675ms
FCS: Checking finished in 14383ms
F# compilation finished in 16534ms
Fable compilation finished in 20158ms

2021-07-03 - Big project
FCS: Parsing finished in 1325ms
FCS: Checking finished in 60134ms
F# compilation finished in 62286ms
Fable compilation finished in 14907ms

2021-11-12 - Standalone
FCS: Parsing finished in 1749ms
FCS: Checking finished in 18818ms
F# compilation finished in 21022ms
Fable compilation finished in 19422ms

2021-11-12 - Big project
FCS: Parsing finished in 1333ms
FCS: Checking finished in 63472ms
F# compilation finished in 65616ms
Fable compilation finished in 14243ms
```

> Note that I've [parallelized the Parsing step](https://github.com/alfonsogarciacaro/fsharp/blob/cf42a5ad4f637966a60da2c4490127130a0e30ef/src/fsharp/FSharp.Compiler.Service/service_slim.fs#L267-L271) in service_slim. This seems to help but doesn't have a big effect overall because TypeCheck is what takes most of the time.

So it looks like the performance decrease doesn't come from the compiler itself but from some bottleneck in service.fs. I'll try to investigate but at least the performance with service_slim is good enough to update Fable to F# 6 :+1:

**Comment by alfonsogarciacaro** [comment]
Scratch the part about parallelizing the parse phase. This can be done in the first compilation, but after that it messes up with cache invalidation so I've removed it from now.

**Comment by MangelMaxime** [comment]
Doing some clean up
